torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -0,0 +1,428 @@
+#include "alias_analysis.h"
+
+#include "torch/csrc/jit/script/error_report.h"
+
+namespace torch {
+namespace jit {
+bool AliasDb::hasWildcard(const Node* n) const {
+  for (const auto input : n->inputs()) {
+    if (valueToAlias_.count(input) != 0 &&
+        valueToAlias_.at(input).isWildcard()) {
+      return true;
+    }
+  }
+
+  for (const auto output : n->outputs()) {
+    if (valueToAlias_.count(output) != 0 &&
+        valueToAlias_.at(output).isWildcard()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool AliasDb::hasWrites(const Node* n) const {
+  for (const auto input : n->inputs()) {
+    if (valueToAlias_.count(input) != 0 && valueToAlias_.at(input).isWrite()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+std::unordered_set<Node*> AliasDb::getWritersForNode(const Node* n) const {
+  // Get all alias sets of this node
+  // ... check the inputs
+  std::unordered_set<Symbol> aliasSets;
+  for (const auto& input : n->inputs()) {
+    if (valueToAlias_.count(input) != 0) {
+      for (const auto& aliasSet : valueToAlias_.at(input).sets()) {
+        aliasSets.insert(aliasSet);
+      }
+    }
+  }
+
+  // ... and the outputs
+  for (const auto& output : n->outputs()) {
+    if (valueToAlias_.count(output) != 0) {
+      for (const auto& aliasSet : valueToAlias_.at(output).sets()) {
+        aliasSets.insert(aliasSet);
+      }
+    }
+  }
+
+  // Then get the union of all writers to all those alias sets
+  std::unordered_set<Node*> writers;
+  for (const auto& alias : aliasSets) {
+    if (aliasToWrites_.count(alias) != 0) {
+      for (const auto writer : aliasToWrites_.at(alias)) {
+        writers.insert(writer);
+      }
+    }
+  }
+  return writers;
+}
+
+void AliasDb::dump() const {
+  std::cout << "\n===1. GRAPH===\n";
+  graph_->dump();
+  std::cout << "===2. ALIAS SETS===\n";
+  for (const auto& pr : valueToAlias_) {
+    std::cout << "%" << pr.first->uniqueName() << " : "
+              << "(";
+
+    bool first = true;
+    for (const auto& alias : pr.second.sets()) {
+      if (first) {
+        first = false;
+      } else {
+        std::cout << ", ";
+      }
+      std::cout << alias.toUnqualString();
+    }
+    std::cout << ")\n";
+  }
+
+  std::cout << "\n===3. WRITES===\n";
+  for (const auto& pr : aliasToWrites_) {
+    std::cout << "Alias set " << pr.first.toUnqualString() << ":\n";
+    for (const auto node : pr.second) {
+      std::cout << "  " << *node;
+    }
+    std::cout << "\n";
+  }
+}
+
+void AliasDb::analyze(std::shared_ptr<Graph> graph) {
+  // Assign aliases to the graph's inputs, assuming that all inputs of a given
+  // type may alias to each other.
+  const auto tensorAlias = getFreshAlias();
+  // Create a separate alias set for each list type
+  std::map<TypeKind, Symbol> listTypeAliases;
```

Comment:
Why are these std::map?

Reply:
`TypeKind` doesn't have a `std::hash` specialization. I'm planning to make one but didn't want to clutter this PR.